lalrpop-test/benches/src/parsers/json.lalrpop

```diff
@@ -0,0 +1,43 @@
+use std::collections::HashMap;
+use std::str::FromStr;
+use crate::src::parsers::json_val::Value;
+
+grammar;
+
+// https://datatracker.ietf.org/doc/html/rfc7159
+
+pub Value: Value = {
+	Object => Value::Object(<>),
+	Array => Value::Array(<>),
+	Number => Value::Num(<>),
+	String => Value::Str(<>),
+	"false" => Value::Boolean(false),
+	"null" => Value::Null,
+	"true" => Value::Boolean(true),
+};
+
+Object: HashMap<String, Value> = {
+	"{" <Comma<Member>> "}" => HashMap::from_iter(<>)
+};
+
+Member: (String, Value) = {
+	<s: String> ":" <v: Value> => (s,v),
+};
+
+Array: Vec<Value> = {
+	"[" <Comma<Value>> "]",
+};
+
+Number: f64 = {
+	r"-?(0|([1-9][0-9]*))(\.[0-9]*)?([eE][-+]?[0-9]+)?" => f64::from_str(<>).unwrap()
+};
+
+String: String = {
+	r#""[^"]*""# => <>.into(),
+};
```

Comment:
Just curious, do other implementations we might benchmark against try to avoid allocating a bunch of new strings here? I could imagine just having a `&'data str` or an arena allocator here. On the other hand, if other parsers typically do this, then its more important to have an apples-to-apples comparison.

Basically, should we be worried about the cost of allocations in this benchmark?

Reply:
"benchmark against" is sort of an interesting term here.  I guess a lot of this depends on our goal.

This file is directly taken from the parser in parse-rosetta-rs, where we are directly compared head to head against other parsers.  In parse-rosetta-rs, glancing through the parsers, it looks like they mostly return allocated `String`s in their data.  Pest seems to use an internal type that takes a lifetime, so may not be allocated, and serde does whatever serde does, but it's a specialized json parser, not a parser generator.  So while it's not universal, we're generally playing by the rules there and having a fair-ish comparison.

Locally, I view these numbers as much more about benchmarking against ourselves, than against any external comparison.  So whatever we do here it will be apples-to-apples unless we're changing the benchmark.  This will allow us to compare PRs that we have performance questions about (eg #952 or #814) without spending a bunch of effort hooking the PR up to an external tool like parse-rosetta-rs.  To that end, the allocation will be apples-to-apples either way.  We might care in terms of allocation dominating performance though.  If allocation does turn out to be very expensive here, then performance changes could appear less impactful than they actually are, causing us to make bad decisions.  I can try to spin up a version of this that skips the allocation and see how expensive it is.  Not sure if I'll have time before the holidays or not, so it may be in 2025.

Reply:
Thanks!

How would you feel about just having both versions as different benchmarks?  I agree that based on those performance numbers the allocation-less version is better, so we definitely want that one.  Having the allocation one as well has two benefits:

1. It lets us know how we'll impact the parse-rosetta-rs metric
2. It keeps a live comparison of the cost of allocations to help answer questions around the cost of allocations going forward.